kart/dataset3_paths.py

```diff
@@ -0,0 +1,285 @@
+        else:
+            raise InvalidOperation(
+                f"This repo uses {encoding!r} path encoding, which isn't supported by this version of Kart"
+            )
+        if levels > 5:
+            raise InvalidOperation(
+                f"This repo uses {levels!r} path levels, which isn't supported by this version of Kart"
+            )
+
+        self.scheme = "int"
+        self.branches = branches
+        self.encoding = encoding
+        self.levels = levels
+        self._mod_value = self.theoretical_max_trees
+
+    def encode_pks_to_path(self, pk_values):
+        assert len(pk_values) == 1
+        pk = pk_values[0]
+        encoded = b64encode_int((pk // self.branches) % self._mod_value)
+        # take chars from the end. so "ABCDE" with two levels becomes "DE"
+        parts = [encoded[i] for i in range(-self.levels, 0)]
+        parts.append(self.encode_filename(pk_values))
+        return "/".join(parts)
+
+    def _nonrecursive_diff(self, tree_a, tree_b):
+        """
+        Returns a dict mapping names to OIDs which differ between the trees.
+        (either the key is present in both, and the OID is different,
+        or the key is only present in one of the trees)
+        """
+        a = {obj.name: obj for obj in tree_a}
+        b = {obj.name: obj for obj in tree_b}
+        all_names = set(a.keys() | b.keys())
+        return {k: (a.get(k), b.get(k)) for k in all_names}
+
+    def max_tree_id(self, repo, base_feature_tree, target_feature_tree):
+        """
+        Looks at a few trees to determine the maximum integer ID of the trees in the given diff.
+        Used as an upper bound for feature count sampling.
+
+        e.g if the only tree is 'A/A/A/A', returns 0
+        """
+        max_tree_path = self._max_feature_tree_path(
+            repo, base_feature_tree, target_feature_tree
+        )
+        return b64decode_int("A" + "".join(max_tree_path.split("/")))
+
+    def _max_feature_tree_path(
+        self, repo, base_feature_tree, target_feature_tree, *, depth=0
+    ):
+        """
+        Returns the path of the tree containing the greatest PK,
+        relative to the given feature tree.
+        """
+        if base_feature_tree == target_feature_tree:
+            return None
+
+        diff = self._nonrecursive_diff(
+            base_feature_tree, target_feature_tree or repo.EMPTY_TREE
+        )
+        max_path = max(diff.keys())
+        if depth == self.levels - 1:
+            return max_path
+        else:
+            a, b = diff[max_path]
+            return (
+                f"{max_path}/{self._max_feature_tree_path(repo, a, b, depth=depth + 1)}"
+            )
+
+    def sample_subtrees(self, num_trees, *, max_tree_id=None):
+        if max_tree_id is None:
+            total_subtrees = self.branches ** self.levels
+        else:
+            total_subtrees = max_tree_id
+        if num_trees >= total_subtrees:
+            # sample all trees
+            yield "."
```

Comment:
It's not invented-by-Craig special since "." does already stand for "the current directory (and implicitly everything in it)" but since every time it's used it has a comment next to it, explaining it, it could benefit from being a constant.

Done